File: Assets/Bundles/Shared/Resources/TS/Util/Signal.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local Object = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local Cancellable = require("Shared/TS/Util/Cancellable").Cancellable
local idCounter = 1
local Signal
do
	Signal = setmetatable({}, {
		__tostring = function()
			return "Signal"
		end,
	})
	Signal.__index = Signal
	function Signal.new(...)
		local self = setmetatable({}, Signal)
		return self:constructor(...) or self
	end
	function Signal:constructor()
		self.debugLogging = false
		self.trackYielding = true
		self.connections = {}
	end
	function Signal:Connect(callback)
		return self:ConnectWithPriority(200, callback)
	end
	function Signal:ConnectWithPriority(priority, callback)
		local id = idCounter
		idCounter += 1
		local item = {
			callback = callback,
			id = id,
		}
		local _connections = self.connections
		local _priority = priority
		if _connections[_priority] ~= nil then
			local _connections_1 = self.connections
			local _priority_1 = priority
			table.insert(_connections_1[_priority_1], item)
		else
			local _connections_1 = self.connections
			local _priority_1 = priority
			local _arg1 = { item }
			_connections_1[_priority_1] = _arg1
		end
		return function()
			local _connections_1 = self.connections
			local _arg0 = function(items)
				local _items = items
				local _arg0_1 = function(item, i)
					if item.id == id then
						local _items_1 = items
						local _i = i
						table.remove(_items_1, _i + 1)
						return nil
					end
				end
				for _k, _v in _items do
					_arg0_1(_v, _k - 1, _items)
				end
			end
			for _k, _v in _connections_1 do
				_arg0(_v, _k, _connections_1)
			end
		end
	end
	function Signal:Once(callback)
		local done = false
		local c
		c = self:Connect(function(...)
			local args = { ... }
			if done then
				return nil
			end
			done = true
			c()
			callback(unpack(args))
		end)
		return c
	end
	function Signal:Fire(...)
		local args = { ... }
		if self.debugLogging then
			-- ▼ ReadonlyMap.size ▼
			local _size = 0
			for _ in self.connections do
				_size += 1
			end
			-- ▲ ReadonlyMap.size ▲
			print("key count: " .. tostring(_size))
			local callbackCount = 0
			for _, priority in Object.keys(self.connections) do
				local _connections = self.connections
				local _priority = priority
				for _1, connection in _connections[_priority] do
					callbackCount += 1
				end
			end
			print("callback count: " .. tostring(callbackCount))
		end
		local fireCount = 0
		local _exp = Object.keys(self.connections)
		local _arg0 = function(a, b)
			return a < b
		end
		table.sort(_exp, _arg0)
		local keys = _exp
		local cancelled = false
		local isCancellable = false
		if #args == 1 and TS.instanceof(args[1], Cancellable) then
			isCancellable = true
		end
		for _, priority in keys do
			local _array = {}
			local _length = #_array
			local _connections = self.connections
			local _priority = priority
			local _array_1 = _connections[_priority]
			table.move(_array_1, 1, #_array_1, _length + 1, _array)
			local entries = _array
			for _1, entry in entries do
				fireCount += 1
				local thread = coroutine.create(entry.callback)
				local success, err = coroutine.resume(thread, unpack(args))
				if not success then
					error(err)
				end
				if coroutine.status(thread) ~= "dead" then
					print(debug:traceback(thread, "Signal yielded unexpectedly. This might be an error."))
				end
				if isCancellable then
					local cancellable = args[1]
					if cancellable:IsCancelled() then
						cancelled = true
						break
					end
				end
			end
			if cancelled then
				break
			end
		end
		if self.debugLogging then
			print("fire count: " .. tostring(fireCount))
		end
		return args[1]
	end
	function Signal:Wait()
		local thread = coroutine.running()
		self:Once(function(...)
			local args = { ... }
			local success, err = coroutine.resume(thread, unpack(args))
			if not success then
				error(err)
			end
		end)
		return coroutine.yield()
	end
	function Signal:Proxy(signal)
		return self:Connect(function(...)
			local args = { ... }
			signal:Fire(unpack(args))
		end)
	end
	function Signal:DisconnectAll()
		table.clear(self.connections)
	end
	function Signal:Destroy()
		self:DisconnectAll()
	end
	function Signal:SetDebug(value)
		self.debugLogging = value
		return self
	end
	function Signal:WithYieldTracking(value)
		self.trackYielding = value
		return self
	end
end
return {
	Signal = Signal,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/@Easy/Core/Core~/src/Shared/Util/Signal.ts
import ObjectUtils from "@easy-games/unity-object-utils";
import { Cancellable } from "./Cancellable";

type SignalParams<T> = Parameters<
	T extends unknown[] ? (...args: T) => never : T extends unknown ? (arg: T) => never : () => never
>;
export type SignalCallback<T> = (...args: SignalParams<T>) => unknown;
type SignalWait<T> = T extends unknown[] ? LuaTuple<T> : T;

interface CallbackItem<T> {
	id: number;
	callback: SignalCallback<T>;
}

export const enum SignalPriority {
	HIGHEST = 0,
	HIGH = 100,
	NORMAL = 200,
	LOW = 300,
	LOWEST = 400,

	/**
	 * The very last priority to get fired.
	 */
	MONITOR = 500,
}

let idCounter = 1;
export class Signal<T extends unknown[] | unknown> {
	private debugLogging = false;
	private trackYielding = true;
	private readonly connections: Map<number, Array<CallbackItem<T>>> = new Map();
	public debugGameObject = false;
	// private readonly connections: Array<CallbackItem<T>> = [];

	/**
	 * Connect a callback function to the signal.
	 *
	 * The returned function can be called to disconnect the callback.
	 */
	public Connect(callback: SignalCallback<T>): () => void {
		return this.ConnectWithPriority(SignalPriority.NORMAL, callback);
	}

	/**
	 * Connect a callback function to the signal.
	 * Highest SignalPriority is called first.
	 *
	 * The returned function can be called to disconnect the callback.
	 */
	public ConnectWithPriority(priority: SignalPriority, callback: SignalCallback<T>): () => void {
		let id = idCounter;
		idCounter++;
		const item: CallbackItem<T> = {
			callback,
			id,
		};
		if (this.connections.has(priority)) {
			this.connections.get(priority)!.push(item);
		} else {
			this.connections.set(priority, [item]);
		}
		return () => {
			for (const [priority, items] of this.connections) {
				let removed = false;
				for (const i of $range(0, items.size() - 1)) {
					const item = items[i];
					if (item.id === id) {
						items.remove(i);
						removed = true;
						break;
					}
				}
				if (removed) {
					if (items.size() === 0) {
						this.connections.delete(priority);
					}
					break;
				}
			}
		};
	}

	/**
	 * Connects a callback function to the signal. The connection is
	 * automatically disconnected after the first invocation.
	 *
	 * The returned function can be called to disconnect the callback.
	 */
	public Once(callback: SignalCallback<T>): () => void {
		let done = false;
		const c = this.Connect((...args) => {
			if (done) return;
			done = true;
			c();
			callback(...args);
		});
		return c;
	}

	/**
	 * Invokes all callback functions with the given arguments.
	 */
	public Fire(...args: SignalParams<T>): T {
		if (this.debugLogging) {
			print("key count: " + this.connections.size());
			let callbackCount = 0;
			for (let priority of ObjectUtils.keys(this.connections)) {
				for (let connection of this.connections.get(priority)!) {
					callbackCount++;
				}
			}
			print("callback count: " + callbackCount);
		}

		let fireCount = 0;
		let keys = ObjectUtils.keys(this.connections).sort((a, b) => a < b);
		let cancelled = false;
		let isCancellable = false;
		if (args.size() === 1 && args[0] instanceof Cancellable) {
			isCancellable = true;
		}
		for (let priority of keys) {
			const entries = [...this.connections.get(priority)!];
			for (let entry of entries) {
				fireCount++;

				const thread = coroutine.create(entry.callback);
				// if (this.debugGameObject) {
				// 	const go = args[0] as GameObject;
				// 	print("fire go.name=" + go.name);
				// }
				const [success, err] = coroutine.resume(thread, ...args);
				if (!success) {
					error(err);
				}
				if (coroutine.status(thread) !== "dead") {
					warn(debug.traceback(thread, "Signal yielded unexpectedly. This might be an error."));
				}

				if (isCancellable) {
					const cancellable = args[0] as Cancellable;
					if (cancellable.IsCancelled()) {
						cancelled = true;
						break;
					}
				}
			}
			if (cancelled) {
				break;
			}
		}
		if (this.debugLogging) {
			print("fire count: " + fireCount);
		}
		return args[0] as T;
	}

	/**
	 * Yields the current thread until the next invocation of the
	 * signal occurs. The invoked arguments will be returned.
	 */
	public Wait() {
		const thread = coroutine.running();
		this.Once((...args) => {
			const [success, err] = coroutine.resume(thread, ...args);
			if (!success) {
				error(err);
			}
		});
		return coroutine.yield() as SignalWait<T>;
	}

	/**
	 * Fires the given signal any time this signal is fired.
	 *
	 * The returned function can be called to disconnect the proxy.
	 */
	public Proxy(signal: Signal<T>) {
		return this.Connect((...args) => {
			signal.Fire(...args);
		});
	}

	/**
	 * Clears all connections.
	 */
	public DisconnectAll() {
		this.connections.clear();
	}

	/**
	 * Returns `true` if there are any connections.
	 */
	public HasConnections() {
		return !this.connections.isEmpty();
	}

	/**
	 * Alias for `DisconnectAll()`.
	 */
	public Destroy() {
		this.DisconnectAll();
	}

	public SetDebug(value: boolean): Signal<T> {
		this.debugLogging = value;
		return this;
	}

	public WithYieldTracking(value: boolean): Signal<T> {
		this.trackYielding = value;
		return this;
	}

	public GetConnectionCount(): number {
		let i = 0;
		for (const value of ObjectUtils.values(this.connections)) {
			i += value.size();
		}
		return i;
	}
}
